File: FileImportDialog.py
from PyQt5.QtWidgets import QDialog
from PyQt5.QtWidgets import QWidget
from PyQt5.QtWidgets import QHBoxLayout
from PyQt5.QtWidgets import QVBoxLayout
from PyQt5.QtWidgets import QPushButton
from PyQt5.QtWidgets import QFileDialog
from PyQt5.QtWidgets import QComboBox
from PyQt5.QtWidgets import QListView
from PyQt5.QtWidgets import QFileSystemModel
from PyQt5.QtWidgets import QLabel
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QFileIconProvider
from PyQt5.QtCore import QSize

import os.path
import sys


__version__ = "0.0.1"


class FileImportDialog(QDialog):

    # define device list
    DEVICES = (
            "Masser Excaliper",
            "Haglöf MDII Caliper",
        )

    FORMATS = (
            "*.txt",
            "*.xml",
        )

    def __init__(self):
        super().__init__()
        #self.setAttribute(Qt.WA_DeleteOnClose)

        # initialize class attributes
        #self._ICON_PATH_32 = os.path.join("resources", "icons", "oxygen", "32")

        self.model = QFileSystemModel()
        self.model.setIconProvider(EmptyIconProvider())
        self.model.setNameFilterDisables(False)

        # setup gui
        #self.setMinimumSize(800, 500)

        # create basic elements
        self.setupUi()

        # connect signals
        self.deviceComboBox.currentIndexChanged.connect(self.updateFileList)
        self.abortImportBtn.clicked.connect(self.reject)
        self.startImportBtn.clicked.connect(self.accept)

        # update file list
        self.updateFileList()

    def setupUi(self):
        self.setWindowTitle("Stichprobe öffnen")

        mainLayout = QVBoxLayout()
        self.setLayout(mainLayout)

        # device selection
        self.deviceLabel = QLabel("Kluppentyp:")
        self.deviceComboBox = QComboBox()
        self.deviceComboBox.setMinimumHeight(60)
        self.deviceComboBox.addItems(self.DEVICES)

        mainLayout.addWidget(self.deviceLabel)
        mainLayout.addWidget(self.deviceComboBox)
        mainLayout.addSpacing(20)

        # file list
        self.fileLabel = QLabel("Stichprobe:")
        self.fileListView = QListView()
        self.fileListView.setMinimumHeight(300)
        self.fileListView.setMinimumWidth(500)
        self.fileListView.setModel(self.model)

        mainLayout.addWidget(self.fileLabel)
        mainLayout.addWidget(self.fileListView)
        mainLayout.addSpacing(20)

        # buttons
        abortIcon = QIcon(os.path.join("resources", "icons", "application-back.svg"))
        self.abortImportBtn = QPushButton(abortIcon, "")
        self.abortImportBtn.setMinimumHeight(60)
        self.abortImportBtn.setIconSize(QSize(40, 40))

        startIcon = QIcon(os.path.join("resources", "icons", "application-next.svg"))
        self.startImportBtn = QPushButton(startIcon, "")
        self.startImportBtn.setMinimumHeight(60)
        self.startImportBtn.setIconSize(QSize(40, 40))
        self.startImportBtn.setDefault(True)
        #self.startImportBtn.setFocus()

        buttonLayout = QHBoxLayout()
        mainLayout.addLayout(buttonLayout)
        buttonLayout.addWidget(self.abortImportBtn)
        buttonLayout.addWidget(self.startImportBtn)

    def updateFileList(self, index=0):
        self.model.setNameFilters((self.FORMATS[index],))
        self.fileListView.setRootIndex(self.model.setRootPath("examples"))

    def getDeviceName(self):
        return self.deviceComboBox.currentText()

    def getFilePath(self):
        return self.model.filePath(self.fileListView.currentIndex())


class EmptyIconProvider(QFileIconProvider):

    def __init__(self):
        super().__init__()

    def icon(self, fileInfo):
        return QIcon()

File: FileImportDialog.cpp
// Copyright (C) 2022 Tobias Helfenstein <tobias@die-softwarezimmerei.de>.
// Licensed under the GPLv3 License. See LICENSE file in the project root for license information.

#include "FileImportDialog.h"

#include <iostream>
#include <QMessagebox>
#include <QtWidgets/QComboBox>
#include <QMessagebox>


FileImportDialog::FileImportDialog() : QDialog()
{
	// create model
	// TODO wird das modell nicht später erst gebraucht?!
	this->model = new QFileSystemModel();
	this->model->setNameFilterDisables(false);

	// create main widget with buttons and set as central widget
	this->setupUi();
	this->updateFileList();
}

FileImportDialog::~FileImportDialog()
{
	// nothing to do
}

void FileImportDialog::setupUi()
{
	this->setWindowTitle("Stichprobe öffnen");

	QVBoxLayout* layout = new QVBoxLayout();
	this->setLayout(layout);

	// device selection
	QLabel* device_lbl = new QLabel("Kluppentyp:");
	this->device_cbx = new QComboBox();
	this->device_cbx->setMinimumHeight(60);
	this->device_cbx->addItem("Masser");
	this->device_cbx->addItem("Haglöf");

	connect(this->device_cbx, QOverload<int>::of(&QComboBox::currentIndexChanged), this, QOverload<int>::of(&FileImportDialog::updateFileList));

	layout->addWidget(device_lbl);
	layout->addWidget(this->device_cbx);
	layout->addSpacing(20);

	// file list
	QLabel* file_lbl = new QLabel("Stichprobe:");
	this->file_lvw = new QListView();
	this->file_lvw->setMinimumHeight(300);
	this->file_lvw->setMinimumWidth(500);
	this->file_lvw->setModel(this->model);

	layout->addWidget(file_lbl);
	layout->addWidget(this->file_lvw);
	layout->addSpacing(20);

	// buttons
	QHBoxLayout* layout_buttons = new QHBoxLayout();
	layout->addLayout(layout_buttons);

	// abort import button
	QIcon icn_abort(":/icons/application-back.svg");
	QPushButton* btn_abort_import = new QPushButton(icn_abort, "");
	btn_abort_import->setMinimumHeight(60);
	btn_abort_import->setIconSize(QSize(40, 40));

	connect(btn_abort_import, &QPushButton::clicked, this, &QDialog::reject);

	layout_buttons->addWidget(btn_abort_import);

	// start import button
	QIcon icn_start(":/icons/application-next.svg");
	QPushButton* btn_start_import = new QPushButton(icn_start, "");
	btn_start_import->setMinimumHeight(60);
	btn_start_import->setIconSize(QSize(40, 40));
	btn_start_import->setDefault(true);

	connect(btn_start_import, &QPushButton::clicked, this, &QDialog::accept);

	layout_buttons->addWidget(btn_start_import);

	return;
}

void FileImportDialog::updateFileList(int index)
{
	QStringList general = {"*.txt", "*.xml"};
	QStringList filters;
	filters << general.at(index).toLocal8Bit().constData();

	this->model->setNameFilters(filters);
	this->file_lvw->setRootIndex(this->model->setRootPath("examples"));

	return;
}

void FileImportDialog::done(int r)
{
	// see https://www.qtcentre.org/threads/8048-Validate-Data-in-QDialog

	if (QDialog::Accepted == r)
	{
		// ok was pressed
		// check for selected file
		if (!this->file_lvw->selectionModel()->hasSelection())
		{
			QMessageBox msg_box;
			msg_box.setWindowTitle("Stichprobenauswertung");
			msg_box.setIcon(QMessageBox::Warning);
			msg_box.setText("Bitte Datei auswählen!");
			msg_box.setStandardButtons(QMessageBox::Ok);
			msg_box.setDefaultButton(QMessageBox::Ok);
			msg_box.exec();

			return;
		}
	}

	QDialog::done(r);

	return;
}

qint64 FileImportDialog::getDeviceIndex()
{
	return this->device_cbx->currentIndex();
}

QString FileImportDialog::getFilePath()
{
	return this->model->filePath(this->file_lvw->currentIndex());
}
